cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -228,139 +258,202 @@ int embed_sim_init(gpi_sim_info_t *info)
-    arg_dict = NULL;
+    pCocotbModule = NULL;
+    pSimLogObject = NULL;
 
     // Ensure that the current thread is ready to call the Python C API
     PyGILState_STATE gstate = PyGILState_Ensure();
     to_python();
 
-    if (get_module_ref(COCOTB_MODULE, &cocotb_module))
+    if (get_module_ref(COCOTB_MODULE, &pCocotbModule))
         goto cleanup;
 
     // Obtain the loggpi logger object
-    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");
+    pSimLogObject = PyObject_GetAttrString(pCocotbModule, "loggpi");    // New reference
 
-    if (simlog_obj == NULL) {
+    if (pSimLogObject == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get simlog object from loggpi\n");
+        goto cleanup;
     }
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "_logFromC");
-    if (simlog_func == NULL) {
+    // Obtain the function to use when logging from C code
+    pLogFn = PyObject_GetAttrString(pSimLogObject, "_logFromC");        // New reference
+    if (pLogFn == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _logFromC method");
         goto cleanup;
     }
-
-    if (!PyCallable_Check(simlog_func)) {
-        PyErr_Print();
+    if (!PyCallable_Check(pLogFn)) {
+        if (PyErr_Occurred())
+            PyErr_Print();
         LOG_ERROR("_logFromC is not callable");
+        Py_DECREF(pLogFn);
         goto cleanup;
     }
 
-    set_log_handler(simlog_func);
+    set_log_handler(pLogFn);                                            // Note: This function steals a reference to pLogFn.
 
-    Py_DECREF(simlog_func);
-
-    simlog_func = PyObject_GetAttrString(simlog_obj, "isEnabledFor");
-    if (simlog_func == NULL) {
+    // Obtain the function to check whether to call log function
+    pFilterFn = PyObject_GetAttrString(pSimLogObject, "isEnabledFor");  // New reference
+    if (pFilterFn == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the isEnabledFor method");
         goto cleanup;
     }
-
-    if (!PyCallable_Check(simlog_func)) {
-        PyErr_Print();
+    if (!PyCallable_Check(pFilterFn)) {
+        if (PyErr_Occurred())
+            PyErr_Print();
         LOG_ERROR("isEnabledFor is not callable");
+        Py_DECREF(pFilterFn);
         goto cleanup;
     }
 
-    set_log_filter(simlog_func);
+    set_log_filter(pFilterFn);                                          // Note: This function steals a reference to pFilterFn.
 
-    argv_list = PyList_New(0);
+    // Build argv for cocotb module
+    pArgvList = PyList_New(info->argc);                                 // New reference
+    if (pArgvList == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Unable to create argv list");
+        goto cleanup;
+    }
     for (i = 0; i < info->argc; i++) {
-        arg_value = PyString_FromString(info->argv[i]);
-        PyList_Append(argv_list, arg_value);
+        pArgvItem = PyString_FromString(info->argv[i]);                 // New reference
```

Comment:
```suggestion
        PyObject *pArgvItem = PyString_FromString(info->argv[i]);                 // New reference
```
and remove the definition above

Reply:
Changed.